File: src/frontend/project/services/auth/supabase.ts
/**
 * Supabase authentication provider implementation.
 */
import { createClient, SupabaseClient } from '@supabase/supabase-js';
import { AuthProvider, AuthSession } from './types';

export class SupabaseAuthProvider implements AuthProvider {
  private client: SupabaseClient;
  private redirectBase: string;

  constructor(
    supabaseUrl: string,
    supabaseKey: string,
    redirectUrl?: string
  ) {
    if (!supabaseUrl || !supabaseKey) {
      throw new Error('Supabase URL and key are required');
    }

    this.client = createClient(supabaseUrl, supabaseKey, {
      auth: {
        autoRefreshToken: true,
        persistSession: true
      }
    });

    this.redirectBase = redirectUrl 
      ? redirectUrl.replace(/\/$/, '') 
      : window.location.origin;
  }

  getProviderName(): string {
    return 'supabase';
  }

  async signInWithGoogle() {
    try {
      const { data, error } = await this.client.auth.signInWithOAuth({
        provider: 'google',
        options: {
          redirectTo: `${this.redirectBase}/auth/callback`,
          queryParams: {
            access_type: 'offline',
          }
        }
      });

      if (error) throw error;
      return { data, error: null };
    } catch (error) {
      console.error('Google sign-in error:', error);
      return { data: null, error };
    }
  }

  async signInWithEmail(email: string, password: string) {
    try {
      const { data, error } = await this.client.auth.signInWithPassword({
        email,
        password,
      });

      if (error) throw error;
      return { data, error: null };
    } catch (error) {
      console.error('Email sign-in error:', error);
      return { data: null, error };
    }
  }

  async signUpWithEmail(email: string, password: string) {
    try {
      const { data, error } = await this.client.auth.signUp({
        email,
        password,
        options: {
          emailRedirectTo: `${this.redirectBase}/auth/callback?noAutoSignIn=true`,
          data: {
            email,
            role: 'free',
            subscription_status: 'none'
          }
        }
      });

      if (error) throw error;

      // Sign out immediately to prevent auto-login
      await this.client.auth.signOut();

      return { data, error: null };
    } catch (error) {
      console.error('Email sign-up error:', error);
      return { data: null, error };
    }
  }

  async signOut() {
    try {
      const { error } = await this.client.auth.signOut();
      if (error) throw error;
    } catch (error) {
      console.error('Sign out error:', error);
      throw error;
    }
  }

  async getSession(): Promise<AuthSession | null> {
    try {
      const { data: { session }, error } = await this.client.auth.getSession();
      if (error) throw error;

      if (!session) return null;

      console.log('[SUPABASE] Raw session:', {
        hasAccessToken: !!session.access_token,
        accessTokenLength: session.access_token?.length,
        hasUser: !!session.user,
        userId: session.user?.id
      });

      return {
        accessToken: session.access_token,
        refreshToken: session.refresh_token,
        tokenType: session.token_type || 'bearer',
        expiresAt: session.expires_at,
        user: session.user ? {
          id: session.user.id,
          email: session.user.email || '',
          emailVerified: !!session.user.email_confirmed_at,
          fullName: session.user.user_metadata?.full_name,
          avatarUrl: session.user.user_metadata?.avatar_url,
          provider: session.user.app_metadata?.provider,
          metadata: session.user.user_metadata
        } : undefined
      };
    } catch (error) {
      console.error('Get session error:', error);
      throw error;
    }
  }

  async resetPassword(email: string) {
    try {
      const { error } = await this.client.auth.resetPasswordForEmail(email, {
        redirectTo: `${this.redirectBase}/auth/callback`,
      });
      if (error) throw error;
    } catch (error) {
      console.error('Reset password error:', error);
      throw error;
    }
  }

  onAuthStateChange(callback: (session: AuthSession | null) => void) {
    const { data: { subscription } } = this.client.auth.onAuthStateChange(
      async (_, session) => {
        if (!session) {
          callback(null);
          return;
        }

        callback({
          accessToken: session.access_token,
          refreshToken: session.refresh_token,
          tokenType: session.token_type || 'bearer',
          expiresAt: session.expires_at,
          user: {
            id: session.user.id,
            email: session.user.email || '',
            emailVerified: !!session.user.email_confirmed_at,
            fullName: session.user.user_metadata?.full_name,
            avatarUrl: session.user.user_metadata?.avatar_url,
            provider: session.user.app_metadata?.provider,
            metadata: session.user.user_metadata
          }
        });
      }
    );

    return { unsubscribe: () => subscription.unsubscribe() };
  }

  async checkUserExists(email: string): Promise<boolean> {
    // Note: This requires direct database access
    // For production, implement via backend API
    console.warn(`checkUserExists should be implemented via backend API for: ${email}`);
    return false;
  }
}

File: src/backend/auth/providers/supabase.py
"""
Supabase authentication provider implementation.

Implements the AuthProvider interface for Supabase Auth.
"""
from typing import Dict, Any, Optional
from datetime import datetime
from supabase import Client, create_client
from ..base import AuthProvider, AuthUser, AuthSession
from src.backend.exceptions import AuthenticationException
from src.backend.utils.logger import setup_logger

logger = setup_logger(__name__)


class SupabaseAuthProvider(AuthProvider):
    """Supabase implementation of AuthProvider interface"""
    
    def __init__(self, url: str, key: str):
        """
        Initialize Supabase auth provider.
        
        Args:
            url: Supabase project URL
            key: Supabase anon/service role key
        """
        if not url or not key:
            raise AuthenticationException(
                "Supabase URL and key are required for Supabase auth provider"
            )
        
        self.client: Client = create_client(url, key)
        logger.info("Initialized Supabase auth provider")
    
    def _map_supabase_user_to_auth_user(self, supabase_user: Any) -> AuthUser:
        """Convert Supabase user object to standardized AuthUser"""
        return AuthUser(
            id=supabase_user.id,
            email=supabase_user.email or "",
            email_verified=supabase_user.email_confirmed_at is not None,
            full_name=supabase_user.user_metadata.get("full_name"),
            avatar_url=supabase_user.user_metadata.get("avatar_url"),
            provider=supabase_user.app_metadata.get("provider", "email"),
            metadata={
                "user_metadata": supabase_user.user_metadata,
                "app_metadata": supabase_user.app_metadata
            }
        )
    
    def _map_supabase_session(self, session: Any) -> AuthSession:
        """Convert Supabase session to standardized AuthSession"""
        return AuthSession(
            access_token=session.access_token,
            refresh_token=session.refresh_token,
            token_type=session.token_type or "bearer",
            expires_at=datetime.fromtimestamp(session.expires_at) if session.expires_at else None,
            user=self._map_supabase_user_to_auth_user(session.user) if session.user else None
        )
    
    async def sign_up(self, email: str, password: str, **kwargs) -> AuthSession:
        """Create new user account with Supabase"""
        try:
            logger.info(f"Sign up attempt for email: {email}")
            
            response = self.client.auth.sign_up({
                "email": email,
                "password": password,
                "options": kwargs.get("options", {})
            })
            
            if not response.session:
                raise AuthenticationException("Sign up failed: No session returned")
            
            logger.info(f"User registered successfully: {response.user.id}")
            return self._map_supabase_session(response.session)
            
        except Exception as e:
            logger.error(f"Sign up failed: {e}", exc_info=True)
            raise AuthenticationException(f"Sign up failed: {str(e)}")
    
    async def sign_in(self, email: str, password: str, **kwargs) -> AuthSession:
        """Authenticate user with email/password"""
        try:
            logger.info(f"Sign in attempt for email: {email}")
            
            response = self.client.auth.sign_in_with_password({
                "email": email,
                "password": password
            })
            
            if not response.session:
                raise AuthenticationException("Sign in failed: No session returned")
            
            logger.info(f"Sign in successful for email: {email}")
            return self._map_supabase_session(response.session)
            
        except Exception as e:
            logger.error(f"Sign in failed for {email}: {e}")
            raise AuthenticationException(f"Sign in failed: {str(e)}")
    
    async def sign_in_with_oauth(
        self, 
        provider: str, 
        redirect_url: str,
        **kwargs
    ) -> Dict[str, Any]:
        """Initiate OAuth sign-in flow"""
        try:
            logger.info(f"OAuth sign-in initiated: provider={provider}")
            
            response = self.client.auth.sign_in_with_oauth({
                "provider": provider,
                "options": {
                    "redirect_to": redirect_url,
                    **kwargs.get("options", {})
                }
            })
            
            return {
                "url": response.url,
                "provider": provider
            }
            
        except Exception as e:
            logger.error(f"OAuth sign-in failed: {e}")
            raise AuthenticationException(f"OAuth sign-in failed: {str(e)}")
    
    async def exchange_code_for_session(self, code: str, **kwargs) -> AuthSession:
        """Exchange OAuth code for session"""
        try:
            logger.debug("Exchanging OAuth code for session")
            
            response = self.client.auth.exchange_code_for_session(code)
            
            if not response.session:
                raise AuthenticationException("Code exchange failed: No session returned")
            
            return self._map_supabase_session(response.session)
            
        except Exception as e:
            logger.error(f"Code exchange failed: {e}")
            raise AuthenticationException(f"Code exchange failed: {str(e)}")
    
    async def verify_token(self, access_token: str) -> AuthUser:
        """Verify access token and return user"""
        try:
            if not access_token:
                raise AuthenticationException("Access token is required")
            
            response = self.client.auth.get_user(access_token)
            
            if not response.user:
                raise AuthenticationException("Invalid or expired token")
            
            return self._map_supabase_user_to_auth_user(response.user)
            
        except AuthenticationException:
            raise
        except Exception as e:
            logger.error(f"Token verification failed: {e}")
            raise AuthenticationException(f"Token verification failed: {str(e)}")
    
    async def refresh_session(self, refresh_token: str) -> AuthSession:
        """Refresh expired access token"""
        try:
            logger.debug("Refreshing session")
            
            response = self.client.auth.refresh_session({
                "refresh_token": refresh_token
            })
            
            if not response.session:
                raise AuthenticationException("Session refresh failed: No session returned")
            
            return self._map_supabase_session(response.session)
            
        except Exception as e:
            logger.error(f"Session refresh failed: {e}")
            raise AuthenticationException(f"Session refresh failed: {str(e)}")
    
    async def sign_out(self, access_token: str) -> bool:
        """Sign out user"""
        try:
            logger.info("User sign out")
            self.client.auth.sign_out()
            return True
        except Exception as e:
            logger.error(f"Sign out failed: {e}")
            return False
    
    async def reset_password(self, email: str, **kwargs) -> bool:
        """Send password reset email"""
        try:
            self.client.auth.reset_password_email(
                email,
                options=kwargs.get("options", {})
            )
            return True
        except Exception as e:
            logger.error(f"Password reset failed: {e}")
            raise AuthenticationException(f"Password reset failed: {str(e)}")
    
    async def update_password(self, access_token: str, new_password: str) -> bool:
        """Update user password"""
        try:
            # Set the token first
            self.client.auth._bearer_token = access_token
            self.client.auth.update_user({"password": new_password})
            return True
        except Exception as e:
            logger.error(f"Password update failed: {e}")
            raise AuthenticationException(f"Password update failed: {str(e)}")
    
    def get_provider_name(self) -> str:
        """Get provider name"""
        return "supabase"
